gui/document.py

```diff
@@ -1531,6 +1531,15 @@ def _brush_settings_changed_cb(self, *a):
             if reset_action.get_sensitive():
                 reset_action.set_sensitive(False)
 
+    def _brush_selected_cb(self, bm, brush, brushinfo):
+        """Internall callback: Go back to the last brush-dependant tool"""
+        # This code assumes the default mode is freehand.
+        # (Or, more accurately, it assumes the default mode is
+        # brush-dependent.)
+        brush_flags = gui.mode.Behavior.PAINT_BRUSH
+        while self.modes.top.pointer_behavior & brush_flags == 0:
+            self.modes.pop()
+
```

Comment:
What about a ModeStack API method called something like `pop_to_behaviour(self, brush_flags)`? To be called by BrushManager._brush_selected_cb().

Might be more understandable. Those `bm, brush, brushinfo` args aren't used here, they're just an artefact of on `@event` declared in some far-off observable.